DDC metrics validation

doasync opened the pull request.

doasync commented:
Can we use a nested HashMap type in Ink? I get the following error:

```zsh
error[E0277]: the trait bound `ink_storage::collections::HashMap<ddc::MetricKey, ddc::MetricValue>: PackedLayout` is not satisfied
  --> lib.rs:40:22
   |
40 |         pub metrics: StorageHashMap<AccountId, StorageHashMap<MetricKey, MetricValue>>,
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `PackedLayout` is not implemented for `ink_storage::collections::HashMap<ddc::MetricKey, ddc::MetricValue>`
   | 
  ::: /Users/doasync/.cargo/registry/src/github.com-1ecc6299db9ec823/ink_storage-3.0.0-rc3/src/collections/hashmap/mod.rs:86:8
   |
86 |     V: PackedLayout,
   |        ------------ required by this bound in `ink_storage::collections::HashMap`
```

Take a look at [this](https://study.impl.dev/hacking/captain-game/#2021-01-01) note. They use `BTreeMap` instead.

doasync commented:
I came up with a new implementation and used an external library to get the median values of the daily metric.

All tests pass and I guess it should work fine, but I have issues with building the smart contract:
```zsh
error: duplicate lang item in crate `std` (which `median` depends on): `oom`.
```

It seems like I just didn't correctly configured this library. @naure, could you please take a look?

naure commented on cere02/lib.rs:
```diff
@@ -642,14 +646,34 @@ mod ddc {
+                let mut filter_bytes = MedianFilter::new(MEDIAN_BUFFER_SIZE);
+                let mut filter_requests = MedianFilter::new(MEDIAN_BUFFER_SIZE);
+
+                for reporter in self.reporters.keys() {
```
A limitation of this is that if we remove or rotate a reporter’s key, its metrics will not be taken into account.

doasync replied:
I assume this is outside the scope of the current task, but what are you suggesting?

naure replied:
No nothing to do right now. Just keep in mind that to remove a reporter also means to ignore its past metrics.

naure commented on cere02/lib.rs:
```diff
@@ -11,6 +11,7 @@ mod ddc {
         lazy::Lazy,
         traits::{PackedLayout, SpreadLayout},
     };
+    use median::Filter as MedianFilter;
```
This library looks quite sophisticated (for a smart contract that is). It’s not impossible but maybe there is a "smaller" way.

doasync replied:
Yeah, I came up with [my own implementation](https://github.com/Cerebellum-Network/cere-enterprise-smart-contracts/pull/22/files#diff-a99b45b57bfdc35f8b28ef154fa4418c459a459aa60dfe0f98d8a350a93fa067R667-R683) of finding the median value. The tests pass the same way as for the `median` library.

naure commented on cere02/lib.rs:
```diff
@@ -612,6 +614,8 @@ mod ddc {
         start_ms: u64,
     }
 
+    const MEDIAN_BUFFER_SIZE: usize = 10;
```
The library "median" is a filter that takes into account only the last X values. Then there is no point in iterating over more X reporters.

It does make sense to limit the number of reporters used for each day. 3 might be good enough. 5 is great. Above may be overkill.

Either way, it’s a small number, so there is no need for a complex data structure.

doasync commented:
@naure Please, take a look at the newest implementation. I used sorting for finding the median value. 

Are we moving in the right direction? Do you have any suggestions for improving [this](https://github.com/Cerebellum-Network/cere-enterprise-smart-contracts/pull/22/files#diff-a99b45b57bfdc35f8b28ef154fa4418c459a459aa60dfe0f98d8a350a93fa067R652-R681) code?

P.S. The tests for metrics validation also look a bit verbose so far. I am going to improve them.

naure commented on cere02/lib.rs:
```diff
@@ -1095,12 +1132,542 @@ mod ddc {
+                MetricValue::default()
+            );
+
+            // bob day1: [0, 6, 8, 8, 100] -> 8
```
Great test!

Maybe also cover the case with just one value?

doasync replied:
[done](https://github.com/Cerebellum-Network/cere-enterprise-smart-contracts/pull/22/files#diff-a99b45b57bfdc35f8b28ef154fa4418c459a459aa60dfe0f98d8a350a93fa067R1216-R1218)

naure reviewed (approved):
See comments.

doasync commented on cere02/Cargo.toml:
```diff
@@ -18,6 +18,8 @@ scale-info = { version = "0.6.0", default-features = false, features = ["derive"
 # Should be removed once bitvecto-rs/bitvec#105 is resolved
 funty = "=1.1.0"
 
+median = "0.3.2"
```
yep, removed

doasync commented on cere02/lib.rs:
```diff
@@ -642,14 +643,46 @@ mod ddc {
                 }
+
+                // vec::sort_unstable is faster, it doesn't preserve the order of equal elements
+                day_stored_bytes.sort_unstable();
```
[done](https://github.com/Cerebellum-Network/cere-enterprise-smart-contracts/pull/22/files#diff-a99b45b57bfdc35f8b28ef154fa4418c459a459aa60dfe0f98d8a350a93fa067R616-R626)

doasync commented:
@naure, I refactored the implementation of finding the median value, moved the logic into a separate function, simplified the code. I also updated the test. So, I guess this functionality is ready... Could you please review/approve?

evgeny-s commented:
@doasync , let's move tests to the separate file
